Чем мы сегодня займёмся? Будем генерировать DnD приключения!

Для начала научимся открывать данные:

In [1]:
path_to_data = 'project/Data/'

In [10]:
path_to_data

'project/Data/'

In [8]:
from tika import parser

In [12]:
raw = parser.from_file(path_to_data+'adv_sacred_temple.pdf')

2019-03-13 15:10:37,037 [MainThread  ] [INFO ]  Retrieving http://search.maven.org/remotecontent?filepath=org/apache/tika/tika-server/1.19/tika-server-1.19.jar to C:\Users\k1l77\AppData\Local\Temp\tika-server.jar.
2019-03-13 15:12:54,422 [MainThread  ] [INFO ]  Retrieving http://search.maven.org/remotecontent?filepath=org/apache/tika/tika-server/1.19/tika-server-1.19.jar.md5 to C:\Users\k1l77\AppData\Local\Temp\tika-server.jar.md5.
2019-03-13 15:12:55,554 [MainThread  ] [WARNI]  Failed to see startup log message; retrying...
2019-03-13 15:13:00,565 [MainThread  ] [WARNI]  Failed to see startup log message; retrying...


In [14]:
print(raw['content'])















































Sacred Temple


OOOPPPEEENNN   GGGAAAMMMIIINNNGGG   
WWWOOORRRLLLDDD   

www.opengamingworld.com

   

   
Sacred Shrine 
A Short Adventure for Four 

2nd Level Characters 
 

Version 1.2 
 

Credits 
Design:   Thomas Leriche 
Play Testers:   Joseph B, Mark H 

 
The Sacred Shrine is a short adventure for 
use with the d20 system by Wizards of the 
Coast®. It requires the use of the Dungeons 
& Dragons® Player’s Handbook, 3.5 
Edition, published by Wizards of the 
Coast®. Dungeons & Dragons® and Wizards 
of the Coast® are Registered Trademarks of 
Wizards of the Coast, and are used with 
Permission under the d20 system license 
(see the LICENSE section). 

 
The Sacred Shrine is a D20 module created for 
the Open Gaming World campaign setting. This 
adventure is for four 2nd level player characters.  
The adventure takes place on and around a 
shrine of Aqual, the elemental god of water. The 
shrine has been abandoned for several months 

d

Текст вычленяется долго, но ладно

Теперь токенизируем его с помощью nltk:

In [15]:
from nltk.tokenize import word_tokenize

In [18]:
len(word_tokenize(raw['content']))

11462

11462 словоупотребления в одном приключении - не так уж и плохо.

Попробуем выкачать все приключения для начинающих игроков:

In [20]:
url = 'http://www.dndadventure.com/html/adventures/adv1.html'

In [22]:
import urllib.request

In [23]:
with urllib.request.urlopen(url) as start_page:
    pagetext = start_page.read().decode('utf-8')

In [25]:
import re

In [27]:
links = re.findall('"http://.*?\.zip"', pagetext)

In [28]:
links

['"http://www.dndadventure.com/ftp/adv_accursed_tower.zip"',
 '"http://www.dndadventure.com/ftp/adv_thief_among.zip"',
 '"http://www.dndadventure.com/ftp/adv_b3palace_of_the_silver_princess.zip"',
 '"http://www.dndadventure.com/ftp/adv_korangar_and_grove.zip"',
 '"http://www.dndadventure.com/ftp/adv_curse_menstaugher_manor.zip"',
 '"http://www.dndadventure.com/ftp/adv_dark_crypt.zip"',
 '"http://www.dndadventure.com/ftp/adv_drestyn1.zip"',
 '"http://www.dndadventure.com/ftp/adv_feast_or_famine.zip"',
 '"http://www.dndadventure.com/ftp/adv_whomwarrages1.zip"',
 '"http://www.dndadventure.com/ftp/adv_gabbersons_keep.zip"',
 '"http://www.dndadventure.com/ftp/adv_ghoulish_surprise.zip"',
 '"http://www.dndadventure.com/ftp/adv_gauntlet.zip"',
 '"http://www.dndadventure.com/ftp/adv_l1_bonehill.zip"',
 '"http://www.dndadventure.com/ftp/adv_l2_knot.zip"',
 '"http://www.dndadventure.com/ftp/adv_last_voy_emer_princess.zip"',
 '"http://www.dndadventure.com/ftp/adv_the_lions_and_the_blades.zip"',
 

In [29]:
import zipfile

In [30]:
help(zipfile)

Help on module zipfile:

NAME
    zipfile - Read and write ZIP files.

DESCRIPTION
    XXX references to utf-8 need further investigation.

CLASSES
    builtins.Exception(builtins.BaseException)
        BadZipFile
        LargeZipFile
    builtins.object
        ZipFile
            PyZipFile
        ZipInfo
    
    class BadZipFile(builtins.Exception)
     |  Common base class for all non-exit exceptions.
     |  
     |  Method resolution order:
     |      BadZipFile
     |      builtins.Exception
     |      builtins.BaseException
     |      builtins.object
     |  
     |  Data descriptors defined here:
     |  
     |  __weakref__
     |      list of weak references to the object (if defined)
     |  
     |  ----------------------------------------------------------------------
     |  Methods inherited from builtins.Exception:
     |  
     |  __init__(self, /, *args, **kwargs)
     |      Initialize self.  See help(type(self)) for accurate signature.
     |  
     |  __new__(

In [41]:
from io import BytesIO

In [39]:
links = list(map(lambda x: x.strip('"'), links))

In [42]:
z = zipfile.ZipFile(BytesIO(urllib.request.urlopen(links[0]).read()))

In [43]:
z.namelist()

['adv_accursed_tower.rtf']

In [44]:
z.close()

Извлечём все документы:

In [45]:
for link in links:
    z = zipfile.ZipFile(BytesIO(urllib.request.urlopen(link).read()))
    z.extractall(path_to_data)
    z.close()

Посмотрим папку, куда извлеклись документы:

In [10]:
import os

In [6]:
[i for i in os.listdir(path_to_data) if i.endswith('.doc')]

['adv_dark_crypt.doc',
 'ForWhomTheWarRages.doc',
 'Hommlet Moathouse.doc',
 'Hommlet NPCs.doc']

У нас есть файлы в формате txt, pdf, doc и rtf. Pdf будем открывать пакетом tika, txt - стандартными средствами питона, rtf - скриптом extract_rtf <a href="https://gist.githubusercontent.com/gilsondev/7c1d2d753ddb522e7bc22511cfb08676/raw/163ef3ee9893109f10cccf4b51d1d077ada296fa/extract_rtf.py">отсюда</a>

DOC извлечь нормально не получится, поэтому тупо конвертим их в RTF при помощи Ворда (тем более, таких файлов всего 4)

In [3]:
import extract_rtf

In [4]:
with open(path_to_data+'Hommlet NPCs.rtf', 'r', encoding='utf-8') as inp:
    rtftext = inp.read()

In [6]:
print(extract_rtf.striprtf(rtftext))

Temple of Elemental Evil NPCs

Elmo’s Father (farmer)
Male Human Fighter 4 CR 4
Hp 20 At +6 (1d8 +2) AC (scale mail + large shield) 17
S 15 I 12 W 16 D 12 Co 16 Ch 11

Elmo
Male Human Ranger 4, CR 4, Medium-sized humanoid (elite)
Hp 40; Init +3; Spd 30 ft; AC 21 (touch 12, flat-footed 19, chainmail +1 and large wooden shield +1) SA Favored enemy (goblinoids + 1), fight with 2 weapons, AL NG
SV Fort +7 Ref +4 Will +4
Atk +10 melee (1d8+5/x3, +1 battleaxe)
Str 19 I 15 W 16 D 16 Co 17 Ch 11
Feats: Power Attack, Weapon Focus (battleaxe), Point Blank shot
(Bonus) Track, Two Weapon Fighting, Ambidexterity
Skills: Animal Empathy +5 Concentration +5 Handle Animal +5 Hide +3 Intuit Direction +5 Knowledge (nature) +6 Listen +6 Move Silently +2 Ride (horse) +9 Search +4 Swim +9 Use Rope +7 Wilderness Lore +7
Spells: delay poison
Possessions: Battleaxe +1, Chainmail +1, Large Wooden Shield +1

Ostler the Innkeeper
Human Male Fighter 2 CR 2
Hp 17 AC 14 (scalemail) At +5 (1d8 + 3)
S 16 I 10 W 13 D 1

In [7]:
texts = []

Извлечём текст из всех документов:

In [11]:
for i in os.listdir(path_to_data):
    if i.endswith('.txt'):
        with open(path_to_data + i, 'r', encoding='utf-8') as inp:
            texts.append(inp.read())
    elif i.endswith('.rtf'):
        with open(path_to_data + i, 'r', encoding='utf-8') as inp:
            texts.append(extract_rtf.striprtf(inp.read()))
    elif i.endswith('.pdf'):
        texts.append(parser.from_file(path_to_data + i)['content'])

2019-03-13 17:19:24,220 [MainThread  ] [WARNI]  Failed to see startup log message; retrying...


In [13]:
len(texts)

27

Всего получилось 27 текстов, посмотрим сколько это токенов:

In [16]:
s = 0
for text in texts:
    s += len(word_tokenize(text))
s

403366

403366 токенов - не так уж и мало, почтиполовина реалека.

Сохраним всё это в json:

In [18]:
import json

In [21]:
with open('project/processed_data.json', 'w', encoding='utf-8') as outp:
    json.dump(texts, outp, ensure_ascii=False)

Посмотрим, сколько у нас символов:

In [22]:
sum(list(map(len, texts)))

1855815

Почти 2 млн символов - так держать!